src/server/pfs/server/driver.go

```diff
@@ -1407,64 +1407,199 @@ func (d *driver) deleteCommit(ctx context.Context, commit *pfs.Commit) error {
 	if err := d.checkIsAuthorized(ctx, commit.Repo, auth.Scope_WRITER); err != nil {
 		return err
 	}
+	// inspect 'commit' to retrieve the commit ID (we'll read it again in a txn)
 	commitInfo, err := d.inspectCommit(ctx, commit, false)
 	if err != nil {
 		return err
 	}
+	commitID := commitInfo.Commit.ID
 
-	if commitInfo.Finished != nil {
-		return fmt.Errorf("cannot delete finished commit")
-	}
+	// Main txn: Delete all downstream commits, and update subvenance of upstream commits
+	deleted := make(map[string]*pfs.CommitInfo, len(commitInfo.Subvenance)) // deleted commits
+	deleteScratch := false                                                  // only delete scratch if txn succeeds
+	if _, err := col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
+		repos := d.repos.ReadWrite(stm)
 
-	// Delete the scratch space for this commit
-	prefix, err := d.scratchCommitPrefix(ctx, commit)
-	if err != nil {
-		return err
-	}
-	_, err = d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix())
-	if err != nil {
-		return err
-	}
+		// re-read CommitInfo inside txn
+		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
+		commitInfo = &pfs.CommitInfo{}
+		if err := commits.Get(commitID, commitInfo); err != nil {
+			return err
+		}
+		deleteScratch = commitInfo.Finished == nil
 
-	// If this commit is the head of a branch, make the commit's parent
-	// the head instead.
-	branchInfos, err := d.listBranch(ctx, commit.Repo)
-	if err != nil {
-		return err
-	}
+		// Helper for deleting commits
+		deleteCommit := func(repo col.ReadWriteCollection, commit *pfs.Commit) error {
```

Comment:
Calling the `ReadWriteCollection` `repo` seems a little bit weird here, I'd just call it `commits`.

Reply:
Done